src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+        } else {
+            ResultSpec::nothing()
+        };
+
+        let error_spec = ResultSpec {
+            fallible: self.fallible,
+            ..Self::nothing()
+        };
+
+        let values_spec = match self.values {
+            Values::Empty => ResultSpec::nothing(),
+            // If this range contains a single datum, just call the function.
+            Values::Within(min, max) if min == max => datum_map(min),
+            // If this is a range of booleans, we know all the values... just try them.
+            Values::Within(Datum::False, Datum::True) => {
+                datum_map(Datum::False).union(datum_map(Datum::True))
+            }
+            // Otherwise, if our function is monotonic, we can try mapping the input
+            // range to an output range.
+            Values::Within(min, max) => match is_monotonic {
+                Monotonic::Yes => {
+                    let min_column = datum_map(min);
+                    let max_column = datum_map(max);
+                    if min_column.nullable
+                        || min_column.fallible
+                        || max_column.nullable
+                        || max_column.fallible
+                    {
+                        ResultSpec::anything()
+                    } else {
+                        min_column.union(max_column)
+                    }
+                }
+                Monotonic::Maybe | Monotonic::No => ResultSpec::anything(),
+            },
+            Values::Nested(_) | Values::All => ResultSpec::anything(),
+        };
+
+        null_spec.union(error_spec).union(values_spec)
+    }
+}
+
+/// [Abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation) for
+/// [MirScalarExpr].
+///
+/// [MirScalarExpr::eval] implements a "concrete interpreter" for the expression type: given an
+/// expression and specific column values as input, it returns a specific value for the output.
+/// This could be reimplemented using this trait... but it's most useful for "abstract"
+/// interpretations of the expression, where we generalize about sets of possible inputs and outputs.
+/// See [Trace] and [ColumnSpecs] for how this can be useful in practice.
+pub trait Interpreter {
+    type Summary: Sized + Clone;
+
+    /// A column of the input row.
+    fn eval_column(&self, id: usize) -> Self::Summary;
+
+    /// A literal value.
+    /// (Stored as a row, because we can't own a Datum.)
+    fn eval_literal(&self, result: &Result<Row, EvalError>, col_type: &ColumnType)
+        -> Self::Summary;
+
+    /// A call to an unmaterializable function.
+    ///
+    /// These functions cannot be evaluated by `MirScalarExpr::eval`. They must
+    /// be transformed away by a higher layer.
+    fn eval_unmaterializable(&self, func: &UnmaterializableFunc) -> Self::Summary;
+
+    /// A function call that takes one expression as an argument.
+    fn eval_unary(&self, func: &UnaryFunc, expr: Self::Summary) -> Self::Summary;
```

Comment:
This interface (and the interface for `eval_binary`, `eval_variadic`, and `eval_if`) all work bottom-up. That seems fine for scalar expressions for the analysis your adding, but other analyses might want to work top-down.

For example, nullability checking can sometimes not bother looking at subterms because it knows the outermost operation won't ever return null.

The savings here are small, but the top-down vs. bottom-up difference will matter more if we scale this framework to include `MirRelationExpr`s, which seems likely. (For example, I'll probably be working on an MIR cost model next, and this framework looks perfect for that!)

Reply:
> [...] but other analyses might want to work top-down.

Assuming I'm understanding the distinction you're making, my first pass at this was a top-down version... the approach in this PR seems to work better for our specific problem, but I can definitely imagine that the top-down approach being more useful in other contexts.

> I'll probably be working on an MIR cost model next, and this framework looks perfect for that!

Neat! Definitely happy to chat about ways we can structure this to make it easier to extend and reuse for future work.